app/domain/invoices/sac_memberships/extend_roles_for_invoicing.rb

```diff
@@ -15,28 +15,90 @@ class ExtendRolesForInvoicing
 
     BATCH_SIZE = 500
 
-    def initialize(date)
-      @date = date
+    def initialize(prolongation_date, reference_date)
+      @prolongation_date = prolongation_date
+      @reference_date = reference_date
     end
 
     def extend_roles
+      convert_roles_to_youth
+      convert_roles_to_adult
+
       roles_to_extend.in_batches(of: BATCH_SIZE) do |batch|
-        Role.with_inactive.where(id: batch.pluck(:id)).update_all(end_on: @date)
+        Role.where(id: batch.pluck(:id)).update_all(end_on: @prolongation_date)
       end
     end
 
     private
 
     def roles_to_extend
-      Role.with_inactive.where(type: ROLES_TO_EXTEND, terminated: false, end_on: ...@date, person_id: person_ids)
+      Role.where(type: ROLES_TO_EXTEND, terminated: false, end_on: ...@prolongation_date, person_id: person_ids)
+    end
+
+    def convert_roles_to_youth
+      roles_to_turn_youth.includes(:group, :person).find_each do |role|
+        if role.person.household.members.count == 2
+          role.person.household.people.each { leave_household(_1) }
+        else
+          leave_household(role.person)
+        end
+      end
+    end
+
+    def convert_roles_to_adult
+      roles_to_turn_adult.includes(:person, :group).find_each do |role|
+        end_role(role)
+        create_mitglied_role(role.person, role.group, :adult)
+
+        role.person.sac_membership.zusatzsektion_roles.where(terminated: false, end_on: ...@prolongation_date, beitragskategorie: :youth).find_each do |zusatzsektion_role|
+          end_role(zusatzsektion_role)
+          create_zusatzmitglied_role(zusatzsektion_role.person, zusatzsektion_role.group, :adult)
+        end
+      end
+    end
+
+    def leave_household(person)
+      Memberships::FamilyMutation.new(person, new_role_end_on: @prolongation_date, new_role_start_on: new_role_start_on, replaced_role_end_on: @reference_date).leave!
+    end
+
+    def roles_to_turn_youth
+      Role.joins(:person).where(type: SacCas::MITGLIED_STAMMSEKTION_ROLES.map(&:sti_name), terminated: false, end_on: ...@prolongation_date, beitragskategorie: :family, person_id: beitragskategorie_change_relevant_person_ids, person: {birthday: turned_youth_reference_age})
+    end
+
+    def roles_to_turn_adult
+      Role.joins(:person).where(type: SacCas::MITGLIED_STAMMSEKTION_ROLES.map(&:sti_name), terminated: false, end_on: ...@prolongation_date, beitragskategorie: :youth, person_id: beitragskategorie_change_relevant_person_ids, person: {birthday: turned_adult_reference_age})
+    end
+
+    def beitragskategorie_change_relevant_person_ids
+      Person.joins(:roles_unscoped)
+        .where(id: person_ids, roles: {end_on: ..@prolongation_date})
+        .where.not(roles: {start_on: @reference_date.., end_on: @prolongation_date..})
```

Comment:
Dazu braucht es unbedingt mehrere Tests. Werden Personen mit diesen Rollen wirklich ausgeschlossen? z.B. sollte sicher der Rollentyp hier noch in die Condition. 

Könnte da auch direkt der `person_ids` Scope erweitert werden? Ich bin irgendwie noch etwas unsicher bezüglich dieser Condition, ob das so wirklich passt...